lib/gamestate.rb

```diff
@@ -0,0 +1,120 @@
+require 'json'
+require_relative 'player'
+require_relative 'room'
+require_relative 'castle'
+
+class Gamestate
+
+  def initialize(castle_data)
+    @castle_data = castle_data
+    @castles = []
+    @castle_order = 0
+    @player = Player.new
+    @castle_data.each do |data|
+      @castles << Castle.new(data)
+    end
+  end
+
+  def current_castle
+    @castles[@castle_order]
+  end
+
+  def current_room
+    current_castle.current_room
+  end
+
+  def next_castle_if_complete
+    if current_castle.complete? && @castle_order < @castles.count - 1
+      @castle_order += 1
+    end
+  end
+
+  def play
+    current_castle.phasing
+    player_move
+    if @player.dead?
+      @player.game_over_report
+      ask_player_to_try_again
+    elsif win_condition?
+      win
+    else
+      play
```

Comment:
Ruby can't recurse indefinitely, so this will crash if you play long enough. Instead, loop forever, play one game inside the loop, and break out of the loop if the player chooses not to play again.